cpp/src/join/mixed_join_semi.cu

```diff
@@ -328,243 +249,37 @@ std::unique_ptr<rmm::device_uvector<size_type>> mixed_join_semi(
-  } else {
-    thrust::counting_iterator<cudf::size_type> stencil(0);
-    auto const [row_bitmask, _] =
-      cudf::detail::bitmask_and(build, stream, rmm::mr::get_current_device_resource());
-    row_is_valid pred{static_cast<bitmask_type const*>(row_bitmask.data())};
-
-    // insert valid rows
-    hash_table.insert_if(
-      iter, iter + right_num_rows, stencil, pred, hash_build, equality_build, stream.value());
-  }
-
-  auto hash_table_view = hash_table.get_device_view();
-
-  // For inner joins we support optimizing the join by launching one thread for
-  // whichever table is larger rather than always using the left table.
-  detail::grid_1d const config(outer_num_rows, DEFAULT_JOIN_BLOCK_SIZE);
-  auto const shmem_size_per_block = parser.shmem_per_thread * config.num_threads_per_block;
-
-  // Allocate storage for the counter used to get the size of the join output
-  rmm::device_scalar<std::size_t> size(0, stream, mr);
-
-  auto const row_hash   = cudf::experimental::row::hash::row_hasher{preprocessed_probe};
-  auto const hash_probe = row_hash.device_hasher(has_nulls);
-
-  // Determine number of output rows without actually building the output to simply
-  // find what the size of the output will be.
-  if (has_nulls) {
-    compute_mixed_join_output_size_semi<DEFAULT_JOIN_BLOCK_SIZE, true>
-      <<<config.num_blocks, config.num_threads_per_block, shmem_size_per_block, stream.value()>>>(
-        *left_conditional_view,
-        *right_conditional_view,
-        *probe_view,
-        *build_view,
-        hash_probe,
-        equality_probe,
-        join_type,
-        hash_table_view,
-        parser.device_expression_data,
-        swap_tables,
-        size.data(),
-        matches_per_row_span);
-  } else {
-    compute_mixed_join_output_size_semi<DEFAULT_JOIN_BLOCK_SIZE, false>
-      <<<config.num_blocks, config.num_threads_per_block, shmem_size_per_block, stream.value()>>>(
-        *left_conditional_view,
-        *right_conditional_view,
-        *probe_view,
-        *build_view,
-        hash_probe,
-        equality_probe,
-        join_type,
-        hash_table_view,
-        parser.device_expression_data,
-        swap_tables,
-        size.data(),
-        matches_per_row_span);
-  }
-
-  return {size.value(stream), std::move(matches_per_row)};
+  auto gather_map = std::make_unique<rmm::device_uvector<size_type>>(probe.num_rows(), stream, mr);
+
+  // gather_map_end will be the end of valid data in gather_map
+  auto gather_map_end =
+    thrust::copy_if(rmm::exec_policy(stream),
+                    thrust::counting_iterator<size_type>(0),
+                    thrust::counting_iterator<size_type>(probe.num_rows()),
+                    flagged->begin(),
+                    gather_map->begin(),
+                    [join_type] __device__(bool flag) {
+                      return flag == (join_type == detail::join_kind::LEFT_SEMI_JOIN);
+                    });
```

Comment:
To go with my other suggestion of renaming `raw_flagged` to `left_table_keep_mask`:
```suggestion
                    [join_type] __device__(bool keep_row) {
                      return keep_row == (join_type == detail::join_kind::LEFT_SEMI_JOIN);
                    });
```